### PR TITLE
Use Kokkos in MemorySpaceData

### PR DESCRIPTION
**Pull request by masterleinad**
This replaces #13549. The idea is really just to replace the pointers in `MemorySpaceData` with `Kokkos::VIew`s which requires more changes in `LinearAlgebra::distributed::Vector`.

Subsequent patches would then replace `MemorySpace::CUDA` with `MemorySpace::Device` in `LinearAlgebra::distributed::Vector` and replace implementation details using CUDA API with `Kokkos` functions.

### CODE REVIEW

File: source/base/kokkos.cc [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        Kokkos::push_finalize_hook(
+          GrowingVectorMemory<
+            LinearAlgebra::distributed::Vector<double, MemorySpace::Device>>::
+            release_unused_memory);
+        Kokkos::push_finalize_hook(
+          GrowingVectorMemory<
+            LinearAlgebra::distributed::Vector<float, MemorySpace::Device>>::
+            release_unused_memory);
+        Kokkos::initialize();
+        std::atexit(Kokkos::finalize);
```
Comment (drwells) [inline source/base/kokkos.cc:57]: Could this be problematic if one of our dependencies uses kokkos and also calls the init/finalize functions?

Comment (masterleinad) [reply source/base/kokkos.cc:57]: Whoever initializes `Kokkos` should also finalize it. If another dependency or user code initializes `Kokkos` before, then they are also responsible for cleaning up and this function does nothing. If this function initializes `Kokkos` no other dependency can initialize `Kokkos` and shouldn't finalize it.

In the end, it's always preferable if user code initializes `Kokkos` but we can't really force that easily (without breaking every code built on top of `deal.II`).

File: source/base/kokkos.cc [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+#ifdef DEAL_II_WITH_CUDA
+        Kokkos::push_finalize_hook(
+          GrowingVectorMemory<
+            LinearAlgebra::distributed::Vector<double, MemorySpace::CUDA>>::
+            release_unused_memory);
+        Kokkos::push_finalize_hook(
+          GrowingVectorMemory<
+            LinearAlgebra::distributed::Vector<float, MemorySpace::CUDA>>::
+            release_unused_memory);
+#endif
```
Comment (masterleinad) [inline source/base/kokkos.cc:59]: We need to make sure that all calls to `Kokkos` functions have happened before `Kokkos` is finalized.

**Comment by masterleinad** [comment]
This requires a newer Trilinos version (also in the CI) that at least includes `Kokkos` 2.6.00. @tjhei said he would update the CI to `Trilinos` 13.2 (`Kokkos` 3.4).

File: include/deal.II/base/memory_space.h [hunk 3]
```diff
@@ -29,15 +33,22 @@ namespace MemorySpace
...
+  {
+    using kokkos_space = ::Kokkos::HostSpace;
+  };
 
+  /**
+   * Structure describing Device memory space.
+   */
+  struct Device
+  {
+    using kokkos_space = ::Kokkos::DefaultExecutionSpace::memory_space;
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/base/memory_space.h:45]: Hmm... I like implementing specializations in terms of `MemorySpace::Host` and `MemorySpace::Device`, and I would rather want to avoid spelling out all possible backends.
Also, leaving it as `DefaultExecutionSpace` also makes the transformation to `Kokkos` simpler using a machine that doesn't have a GPU. It also makes it easier to compare the current `Host` implementations with a `Kokkos` implementation on the host.

What do others think?

[the remark this answers is not in the document]
Comment (Rombur) [reply include/deal.II/base/memory_space.h:45]: I am fine using `DefaultExecutsionSpace` but it shouldn't be called `Device`. Right now, `Host` means CPU and `Device` means CPU or GPU. If I use `MemorySpace::Device` on a machine that has a GPU, I would expect that my code to use the GPU or to crash if it can't. I would not expect it to run on the CPU.

[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/base/memory_space.h:45]: I changed it to `MemorySpace::Default`.

File: include/deal.II/base/memory_space_data.h [hunk 4]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
...
-    {
-      (void)begin;
-      (void)n_elements;
-    }
+    copy_from(const T *begin, const std::size_t n_elements);
+
+    /**
+     * Kokkos View to a host buffer used for MPI communication
+     */
+    Kokkos::View<T *, Kokkos::HostSpace> values_host_buffer;
```
Comment (Rombur) [inline include/deal.II/base/memory_space_data.h:67]: Is this still necessary? Before we needed this because memory allocation was different for `values` and `values_dev`. IIRC we used `values` as the host buffer for MPI because we wanted to reuse the `Host` code that used `values`. I think now we don't want this anymore and instead we should use `create_mirror_view`. We don't need to treat this in the PR but then we need a FIXME

File: include/deal.II/lac/la_parallel_vector.templates.h [hunk 5]
```diff
@@ -925,21 +930,27 @@ namespace LinearAlgebra
...
+            }
+          else
 #endif
+            {
+              std::fill_n(data.values.data() +
+                            partitioner->locally_owned_size(),
+                          partitioner->n_ghost_indices(),
+                          Number());
+            }
+        }
```
Comment (Rombur) [inline include/deal.II/lac/la_parallel_vector.templates.h:953]: All of this can be cleaned up but we can do it in a different PR.

File: include/deal.II/lac/la_parallel_vector.templates.h [hunk 6]
```diff
@@ -1001,34 +1005,19 @@ namespace LinearAlgebra
...
-        }
-#  endif
-
-#  if defined(DEAL_II_COMPILER_CUDA_AWARE) && \
-    defined(DEAL_II_MPI_WITH_CUDA_SUPPORT)
-      if (std::is_same<MemorySpaceType, dealii::MemorySpace::CUDA>::value)
-        {
+          data.values_host_buffer =
+            Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{},
+                                                data.values);
```
Comment (Rombur) [inline include/deal.II/lac/la_parallel_vector.templates.h:1010]: Don't you need to fence `HostSpace` before using `data.values_host_buffer`?

File: include/deal.II/lac/la_parallel_vector.templates.h [hunk 7]
```diff
@@ -1146,77 +1133,67 @@ namespace LinearAlgebra
...
-#  endif
+      if (std::is_same<MemorySpaceType, MemorySpace::CUDA>::value)
+        {
+          // Move the data to the host and then move it back to the
+          // device. We use values to store the elements because the function
+          // uses a view of the array and thus we need the data on the host to
+          // outlive the scope of the function.
+          data.values_host_buffer =
+            Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{},
+                                                data.values);
```
Comment (Rombur) [inline include/deal.II/lac/la_parallel_vector.templates.h:1167]: No fence?

File: include/deal.II/lac/vector_memory.templates.h [hunk 8]
```diff
@@ -32,7 +33,21 @@ template <typename VectorType>
...
 {
-  static GrowingVectorMemory<VectorType>::Pool pool;
+  // Kokkos needs to be initialized before constructing the static Pool for
+  // vector types that use Kokkos.
+  // If Kokkos is initialized by deal.II, this make sure that it is finalized
+  // after the Pool has been destroyed.
+  // If Kokkos is not initialized by deal.II, we assume that Kokkos is not
+  // finalized past program end together with static variables and we need to
+  // make sure to empty the Pool when finalizing Kokkos so that the destruction
+  // of the Pool doesn't call Kokkos functions.
```
Comment (Rombur) [inline include/deal.II/lac/vector_memory.templates.h:43]: I hate `GrowingVectorMemory` so much...

File: include/deal.II/lac/vector_operations_internal.h [hunk 9]
```diff
@@ -2112,8 +2116,8 @@ namespace internal
       {
         if (operation == VectorOperation::insert)
           {
-            cudaError_t cuda_error_code = cudaMemcpy(data.values.get(),
-                                                     v_data.values_dev.get(),
+            cudaError_t cuda_error_code = cudaMemcpy(data.values.data(),
+                                                     v_data.values.data(),
```
Comment (Rombur) [inline include/deal.II/lac/vector_operations_internal.h:2120]: It's going to be nice to clean up all of that

File: include/deal.II/base/memory_space_data.h [hunk 10]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
...
-    {
-      static_assert(std::is_same<MemorySpace, Host>::value ||
-                      std::is_same<MemorySpace, CUDA>::value,
-                    "MemorySpace should be Host or CUDA");
-    }
+    MemorySpaceData();
 
     /**
-     * Copy the active data (values for Host and values_dev for CUDA) to @p begin.
+     * Copy the active data (values for Host and values_dev for Device) to @p begin.
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h:51]: Done.

File: include/deal.II/base/memory_space_data.h [hunk 11]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
...
-    {
-      (void)begin;
-      (void)n_elements;
-    }
+    copy_to(T *begin, const std::size_t n_elements);
 
     /**
      * Copy the data in @p begin to the active data of the structure (values for
-     * Host and values_dev for CUDA). The pointer @p begin must be on the host.
+     * Host and values_dev for Device). The pointer @p begin must be on the host.
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h:59]: Done.

File: include/deal.II/base/memory_space_data.h (hunk 4 above)
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h:67]: We need to store the buffer somewhere to survive function scope where we could use `create_mirror_view` to create a local buffer. I added a FIXME.

File: include/deal.II/base/memory_space_data.h [hunk 12]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
...
-     * Pointer to data on the device.
+     * Pointer to data on the host. The pointer points to the same data as
+     * values when using shared memory. Otherwise it is not set.
      */
-    std::unique_ptr<Number[]> values_dev;
+    // The pointer is shared so that MemorySpaceData can be copied and
+    // MemorySpaceData::values can be used in Kokkos::parallel_for. This
+    // pointer owns the data when using shared memory with MPI. In this case,
+    // the Kokkos::View in non-owning. When shared memory with MPI is not used,
+    // the pointer is not used.
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h:82]: No, this is only ever used with `MemorySpace::Host`. I added a comment.

File: include/deal.II/lac/la_parallel_vector.templates.h (hunk 5 above)
Comment (masterleinad) [reply include/deal.II/lac/la_parallel_vector.templates.h:953]: Yes, see https://github.com/dealii/dealii/pull/14537/files#diff-f2ee1054edd7dc30feb86f7331a47a9a0c50f82de59bb66c61d8de73f062a96f.

File: include/deal.II/lac/la_parallel_vector.templates.h (hunk 6 above)
Comment (masterleinad) [reply include/deal.II/lac/la_parallel_vector.templates.h:1010]: No, we don't pass an execution space. So the function has a fence.

File: include/deal.II/lac/la_parallel_vector.templates.h (hunk 7 above)
Comment (masterleinad) [reply include/deal.II/lac/la_parallel_vector.templates.h:1167]: We don't pass an execution space. So the function has a fence.

File: include/deal.II/lac/vector_operations_internal.h (hunk 9 above)
Comment (masterleinad) [reply include/deal.II/lac/vector_operations_internal.h:2120]: Yes, see https://github.com/dealii/dealii/pull/14537/files#diff-acb1ab2a77fc1812186da8b93c1e717f14af0e93dee3b2535efe906aa67c7c95.

**Review by Rombur — APPROVED** [review]

File: source/base/kokkos.cc [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+#include <deal.II/base/kokkos.h>
+
+#include <deal.II/lac/la_parallel_vector.h>
+#include <deal.II/lac/vector_memory.h>
+
+#include <Kokkos_Core.hpp>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Impl
```
Comment (peterrum) [inline source/base/kokkos.cc:25]:
```suggestion
namespace internal
```
? Is there any reason to use the name `Impl`?

File: include/deal.II/base/kokkos.h [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+// ---------------------------------------------------------------------
+
+#ifndef dealii_kokkos_h
+#define dealii_kokkos_h
+
+#include <deal.II/base/config.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Impl
```
Comment (peterrum) [inline include/deal.II/base/kokkos.h:23]: What about
```suggestion
namespace Kokkos
```
or ``KokkosWrappers`?

File: include/deal.II/lac/la_parallel_vector.templates.h [hunk 15]
```diff
@@ -135,17 +135,12 @@ namespace LinearAlgebra
         {
           if (comm_shared == MPI_COMM_SELF)
             {
-              Number *new_val;
-              Utilities::System::posix_memalign(
-                reinterpret_cast<void **>(&new_val),
-                64,
-                sizeof(Number) * new_alloc_size);
-              data.values = {new_val, [](Number *data) { std::free(data); }};
+              Kokkos::resize(data.values, new_alloc_size);
```
Comment (peterrum) [inline include/deal.II/lac/la_parallel_vector.templates.h:138]: Where is alignment specified?

File: include/deal.II/lac/la_parallel_vector.templates.h [hunk 16]
```diff
@@ -229,13 +224,25 @@ namespace LinearAlgebra
...
+              // that.
+              data.values_sm_ptr = {ptr_aligned,
+                                    [mpi_window](Number *) mutable {
+                                      // note: we are creating here a copy of
+                                      // the window other approaches led to
+                                      // segmentation faults
+                                      const auto ierr =
+                                        MPI_Win_free(&mpi_window);
+                                      AssertThrowMPI(ierr);
+                                    }};
```
Comment (peterrum) [inline include/deal.II/lac/la_parallel_vector.templates.h:244]: :+1:

File: include/deal.II/base/memory_space_data.h [hunk 17]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
...
 
-    std::vector<ArrayView<const Number>> values_sm;
-  };
+  template <typename T, typename MemorySpace>
+  MemorySpaceData<T, MemorySpace>::MemorySpaceData()
+    : values_host_buffer(
+        (dealii::Impl::ensure_kokkos_initialized(),
+         Kokkos::View<T *, Kokkos::HostSpace>("host buffer", 0)))
+    , values(Kokkos::View<T *, typename MemorySpace::kokkos_space>(
+        "memoryspace data",
```
Comment (peterrum) [inline include/deal.II/base/memory_space_data.h:109]: Is Kokkos using strings!?

File: include/deal.II/base/memory_space_data.h [hunk 18]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
...
+     */
+    // FIXME Should we move this somewhere else?
+    Kokkos::View<T *, Kokkos::HostSpace> values_host_buffer;
 
     /**
-     * Pointer to data on the host.
+     * Kokkos View to the data on the device
      */
-    std::unique_ptr<Number[], std::function<void(Number *)>> values;
+    Kokkos::View<T *, typename MemorySpace::kokkos_space> values;
```
Comment (peterrum) [inline include/deal.II/base/memory_space_data.h:71]: Is this a view or does it also contain the actual values? I am confused by the name and how it used.

**Review by peterrum — COMMENTED** [review]
Looks good! I have a few questions. 

Have you run some matrix-free benchmarks? I guess if Kokkos maps to the same `posix_memalign` functions we are currently using the performance should be the same.

File: include/deal.II/base/memory_space_data.h (hunk 18 above)
Comment (Rombur) [reply include/deal.II/base/memory_space_data.h:71]: Kokkos::View is owning unless you use an unmanaged View.  A Kokkos::View is a really some kind of shared pointer + metadata

File: include/deal.II/lac/la_parallel_vector.templates.h (hunk 15 above)
Comment (Rombur) [reply include/deal.II/lac/la_parallel_vector.templates.h:138]: It's a macro in Kokkos. The default value is 64

File: include/deal.II/base/kokkos.h (hunk 14 above)
Comment (masterleinad) [reply include/deal.II/base/kokkos.h:23]: At this point, I don't see a reason for this to be user-facing.

File: include/deal.II/base/memory_space_data.h (hunk 17 above)
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h:109]: Yes, managed Views store a label that is used, e.g., in `Kokkos` profiling tools.

File: source/base/kokkos.cc (hunk 13 above)
Comment (masterleinad) [reply source/base/kokkos.cc:25]: That's a common namespace in `Kokkos`. I'm happy to change it to `internal`.

File: include/deal.II/base/memory_space_data.h (hunk 18 above)
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h:71]: The number of `*`s in the first template argument gives you the dimensionality of the `View`. So far we are only using one-dimensional Views. The second template argument is the memory space. You could specify more template arguments nut they are not really needed in general.

File: include/deal.II/lac/la_parallel_vector.templates.h (hunk 15 above)
Comment (masterleinad) [reply include/deal.II/lac/la_parallel_vector.templates.h:138]: See https://github.com/kokkos/kokkos/blob/5ad609661e570ba6aa7716a26a91cb67d559f8a2/core/src/Kokkos_Macros.hpp#L241-L243.

**Review by peterrum — APPROVED** [review]
Thanks for the answers. Let's rename the namespace. After that the PR is fine from my side!

File: include/deal.II/lac/vector_memory.templates.h (hunk 8 above)
Comment (kronbichler) [reply include/deal.II/lac/vector_memory.templates.h:43]: I agree, that class is a bit weird. I'd rather prefer to explicitly manage pools at the places they make sense. (I've written my own `VectorMemory` classes several times to make them less intrusive than the `GrowingVectorMemory`.

**Review by kronbichler — APPROVED** [review]

File: include/deal.II/base/memory_space_data.h [hunk 19]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
...
 
     /**
-     * Pointer to data on the device.
+     * Pointer to data on the host. The pointer points to the same data as
+     * values when using shared memory and the memory space is
+     * MemorySpace::Host. Otherwise it is not set.
      */
-    std::unique_ptr<Number[]> values_dev;
+    // The pointer is shared so that MemorySpaceData can be copied and
```
Comment (tjhei) [inline include/deal.II/base/memory_space_data.h:78]: This is a shared pointer so that ...

File: include/deal.II/base/memory_space_data.h [hunk 20]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
...
+    // MemorySpaceData::values can be used in Kokkos::parallel_for. This
+    // pointer owns the data when using shared memory with MPI. In this case,
+    // the Kokkos::View in non-owning. When shared memory with MPI is not used,
```
Comment (tjhei) [inline include/deal.II/base/memory_space_data.h:81]:
```suggestion
    // the Kokkos::View is non-owning. When shared memory with MPI is not used,
```
Which view, @p values?

File: include/deal.II/base/memory_space_data.h [hunk 21]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
...
+    // the pointer is not used.
```
Comment (tjhei) [inline include/deal.II/base/memory_space_data.h:82]: Which pointer?

File: include/deal.II/base/memory_space_data.h [hunk 22]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
...
-  inline void
-  swap(MemorySpaceData<Number, Host> &u, MemorySpaceData<Number, Host> &v)
+  template <typename T, typename MemorySpace>
+  void
+  MemorySpaceData<T, MemorySpace>::copy_to(T *               begin,
+                                           const std::size_t n_elements)
   {
-    std::swap(u.values, v.values);
+    Assert(n_elements <= values.extent(0),
+           ExcMessage("n_elements greater than the size of values."));
```
Comment (tjhei) [inline include/deal.II/base/memory_space_data.h:121]:
```suggestion
           ExcMessage("n_elements is greater than the size of MemorySpaceData."));
```

File: include/deal.II/base/memory_space_data.h [hunk 23]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
...
-    std::unique_ptr<Number[], std::function<void(Number *)>> values;
-    std::unique_ptr<Number[], void (*)(Number *)>            values_dev;
-
-    /**
-     * This is currently not used.
-     */
-    std::vector<ArrayView<const Number>> values_sm;
-  };
+    Assert(n_elements <= values.extent(0),
+           ExcMessage("n_elements greater than the size of values."));
```
Comment (tjhei) [inline include/deal.II/base/memory_space_data.h:141]:
```suggestion
           ExcMessage("n_elements is greater than the size of values."));
```

File: include/deal.II/base/memory_space.h [hunk 24]
```diff
@@ -29,15 +33,23 @@ namespace MemorySpace
...
+  };
 
   /**
    * Structure describing CUDA memory space.
    */
-  struct CUDA
-  {};
+  // FIXME Only enable if CUDA is enabled in deal.II.
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/base/memory_space.h:54]: Yes, at the moment we use `CUDA` explicitly in a lot of places that I would not touch otherwise. That's why I'd rather do it in a separate pull request.

File: include/deal.II/base/memory_space_data.h [hunk 25]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
 namespace MemorySpace
 {
   /**
-   * Data structure
+   * Structure which stores data on the host or the device depending on the
+   * template parameter @p MemorySpace. The data is copied into the structure
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h]: Sure.

File: include/deal.II/base/memory_space.h [hunk 26]
```diff
@@ -29,15 +33,23 @@ namespace MemorySpace
...
    */
   struct Host
-  {};
-
+  {
+    using kokkos_space = ::Kokkos::HostSpace;
+  };
 
+  /**
+   * Structure describing the default memory space.
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/base/memory_space.h]: Sure.

File: include/deal.II/base/memory_space_data.h (hunk 18 above)
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h:74]: Sure.

File: include/deal.II/base/memory_space_data.h [hunk 27]
```diff
@@ -32,165 +37,134 @@ DEAL_II_NAMESPACE_OPEN
...
-     * Pointer to data on the host.
+     * Kokkos View to the data on the device
      */
-    std::unique_ptr<Number[], std::function<void(Number *)>> values;
+    Kokkos::View<T *, typename MemorySpace::kokkos_space> values;
 
     /**
-     * Pointer to data on the device.
+     * Pointer to data on the host. The pointer points to the same data as
+     * values when using shared memory and the memory space is
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h]: Done.

File: include/deal.II/base/memory_space_data.h (hunk 19 above)
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h]: Done.

File: include/deal.II/base/memory_space_data.h (hunk 20 above)
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h]: Yes, done.

File: include/deal.II/base/memory_space_data.h (hunk 21 above)
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h]: `values_sm_ptr`. Done.

File: include/deal.II/base/memory_space_data.h (hunk 22 above)
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h]: Done.

File: include/deal.II/base/memory_space_data.h (hunk 23 above)
Comment (masterleinad) [reply include/deal.II/base/memory_space_data.h]: Done.

**Review by drwells — APPROVED** [review]
This looks good - could you rebase before we merge?

**Review by kronbichler — APPROVED** [review]

File: include/deal.II/base/memory_space.h [hunk 28]
```diff
@@ -29,15 +33,26 @@ namespace MemorySpace
...
    */
   struct Host
-  {};
-
+  {
+    using kokkos_space = ::Kokkos::HostSpace;
+  };
 
+  /**
+   * Structure describing the default memory space. If Kokkos was configure with
```
Comment (tjhei) [inline include/deal.II/base/memory_space.h:41]:
```suggestion
   * Structure describing the default memory space. If Kokkos was configured with
```

**Review by tjhei — APPROVED** [review]

File: include/deal.II/base/memory_space.h (hunk 28 above)
Comment (tjhei) [reply include/deal.II/base/memory_space.h:41]: Found one more typo, sorry @masterleinad .
